### PR TITLE
WIP: Preact EX

### PR DESCRIPTION
**Pull request by theKashey**
Full create preact integration

Spiking injection into the deepest deeps of Preact to make RHL work without VDOM and hot-render.

### CODE REVIEW

File: examples/preact/src/App.js [hunk 1]
```diff
@@ -3,12 +3,23 @@ import { hot } from 'react-hot-loader'
 import Counter from './Counter'
 import { Internal } from './Internal'
 
+const Inner = ({ children }) => <div>{children}</div>
+
+const indirect = {
+  element: () => (
```
Comment (theKashey) [inline examples/preact/src/App.js]: This element is "invisible" to RHL "v3". Ie for RHL+Preact+(no-VDOM)
The internal state always will be lost.

File: examples/preact/src/hotLoaderSetup.js [hunk 2]
```diff
@@ -1,4 +1,34 @@
-import reactHotLoader from 'react-hot-loader'
-import preact from 'preact'
+import reactHotLoader, { compareOrSwap } from 'react-hot-loader'
+import preact, { setComponentComparator } from 'preact'
 
 reactHotLoader.inject(preact, 'h')
+
+setComponentComparator(compareOrSwap)
```
Comment (theKashey) [inline examples/preact/src/hotLoaderSetup.js]: settingComparator to compareAndSwap (swap is optional, so - "or")

File: examples/preact/src/hotLoaderSetup.js [hunk 3]
```diff
@@ -1,4 +1,34 @@
...
+//(oldC, newC) => compareOrSwap(oldC, Object.getPrototypeOf(newC)));
+
+// changes to preact
+
```
Comment (theKashey) [inline examples/preact/src/hotLoaderSetup.js]: This is changes to be added into Preact code base. Literraly - just wrapping `node._componentConstructor == vnode.nodeName` by a controllable function.

File: src/proxy/createClassProxy.js [hunk 4]
```diff
@@ -225,40 +225,47 @@ function createClassProxy(InitialComponent, proxyKey, options) {
...
-      // style wrapping for this as we do elsewhere, so just we just pass it
-      // through as-is.
-      if (isReactComponentInstance(result)) {
-        ProxyComponent = null
-        return result
-      }
-
-      // Otherwise, it's a normal functional component. Build the real proxy
-      // and use it going forward.
+    if (1) {
```
Comment (theKashey) [inline src/proxy/createClassProxy.js]: SFC have to be wrapped by "Components", or Preact will wrap them. (to be controlled via setConfig)

File: src/reconciler/hotReplacementRender.js [hunk 5]
```diff
@@ -58,7 +58,7 @@ const haveTextSimilarity = (a, b) =>
 
 const equalClasses = (a, b) => {
   const prototypeA = a.prototype
-  const prototypeB = Object.getPrototypeOf(b.prototype)
+  const prototypeB = b.prototype // Object.getPrototypeOf(b.prototype)
```
Comment (theKashey) [inline src/reconciler/hotReplacementRender.js]: In React we are comparing unwrapped class with wrapped (real is behind getPrototypeOf). Should be controllable.

File: src/reconciler/hotReplacementRender.js [hunk 6]
```diff
@@ -348,6 +348,21 @@ const hotReplacementRender = (instance, stack) => {
   }
 }
 
+export const hotComponentCompare = (oldType, newType) => {
+  if (oldType === newType) {
+    return true
+  }
+
+  if (isSwappable(newType, oldType)) {
```
Comment (theKashey) [inline src/reconciler/hotReplacementRender.js]: This is the actual change.
If type == type - return.
Else - compare them. Both types are "wrapped".
If types are replaceable
 - rollback proxy1
 - set alias oldProxy->newProxy
 - set alias newProxy(usedId) -> oldProxy (may be obsolete)

return true.

Result it 💯 full cream hot-loader.

File: examples/preact/src/hotLoaderSetup.js [hunk 7]
```diff
@@ -1,4 +1,34 @@
-import reactHotLoader from 'react-hot-loader'
-import preact from 'preact'
+import reactHotLoader, { compareOrSwap } from 'react-hot-loader'
+import preact, { setComponentComparator } from 'preact'
 
 reactHotLoader.inject(preact, 'h')
```
Comment (developit) [inline examples/preact/src/hotLoaderSetup.js]: better to use `preact.options.vnode` here.

**Pull request closed without merging**